examples/chef/common/chef-fan-control-manager.cpp

```diff
@@ -0,0 +1,178 @@
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterface.h>
+#include <app/clusters/fan-control-server/fan-control-server.h>
+#include <app/util/attribute-storage.h>
+#include <app/util/error-mapping.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::FanControl;
+using namespace chip::app::Clusters::FanControl::Attributes;
+using Protocols::InteractionModel::Status;
+
+namespace {
+class ChefFanControlManager : public AttributeAccessInterface, public Delegate
+{
+public:
+    // Register for the FanControl cluster on all endpoints.
+    ChefFanControlManager(EndpointId aEndpointId) :
+        AttributeAccessInterface(Optional<EndpointId>(aEndpointId), FanControl::Id), Delegate(aEndpointId)
+    {}
+
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+    Status HandleStep(StepDirectionEnum aDirection, bool aWrap, bool aLowestOff) override;
+
+private:
+    CHIP_ERROR ReadPercentCurrent(AttributeValueEncoder & aEncoder);
+    CHIP_ERROR ReadSpeedCurrent(AttributeValueEncoder & aEncoder);
+};
+
+static ChefFanControlManager * mFanControlManager = nullptr;
+
+CHIP_ERROR ChefFanControlManager::ReadPercentCurrent(AttributeValueEncoder & aEncoder)
+{
+    // Return PercentSetting attribute value for now
+    DataModel::Nullable<Percent> percentSetting;
+    PercentSetting::Get(mEndpoint, percentSetting);
+    Percent ret = 0;
+    if (!percentSetting.IsNull())
+    {
+        ret = percentSetting.Value();
+    }
+
+    return aEncoder.Encode(ret);
+}
+
+CHIP_ERROR ChefFanControlManager::ReadSpeedCurrent(AttributeValueEncoder & aEncoder)
+{
+    // Return SpeedCurrent attribute value for now
+    DataModel::Nullable<uint8_t> speedSetting;
+    SpeedSetting::Get(mEndpoint, speedSetting);
+    uint8_t ret = 0;
+    if (!speedSetting.IsNull())
+    {
+        ret = speedSetting.Value();
+    }
+
+    return aEncoder.Encode(ret);
+}
+
+Status ChefFanControlManager::HandleStep(StepDirectionEnum aDirection, bool aWrap, bool aLowestOff)
+{
+    ChipLogProgress(NotSpecified, "ChefFanControlManager::HandleStep aDirection %d, aWrap %d, aLowestOff %d",
+                    to_underlying(aDirection), aWrap, aLowestOff);
+
+    VerifyOrReturnError(aDirection != StepDirectionEnum::kUnknownEnumValue, Status::InvalidCommand);
+
+    uint8_t speedMax;
+    SpeedMax::Get(mEndpoint, &speedMax);
+
+    DataModel::Nullable<uint8_t> speedSetting;
+    SpeedSetting::Get(mEndpoint, speedSetting);
+
+    uint8_t newSpeedSetting = speedSetting.IsNull() ? 0 : speedSetting.Value();
+
+    if (aDirection == StepDirectionEnum::kIncrease)
+    {
+        if (speedSetting.IsNull())
+        {
+            newSpeedSetting = 1;
```

Comment:
this should probably use aLowestOff instead of 1 (so that increase from null is 1 if aLowestOff is 0.

Reply:
Done. PTAL. Thanks!

Reply:
I do not see changes.

Reply:
Adding kLowest  done. PTAL.  
BTW, when speedSetting.IsNull()  the newSpeedSetting should be 1 instead of kLowest.

Reply:
```cpp
newSpeedSetting = std::invoke([&]{
   VerifyOrReturn(!speedSetting.IsNull(), kLowest);

   if (speedSetting.Value() < speedMax) 
   {
      return static_cast<uint8_t>(speedSetting.Value() + 1);
   }
   
   return aWrap ? kLowest : speedSetting.Value();
});
```

Reply:
If `aLowestOff` is true and `speedSetting.IsNull()` , does it make sense to set it to 0 when increasing speed?

Reply:
It is unclear to me what the intended behaviour is. I just assumed this was what we tried to do for symmetry. If setting to 1 makes sense (so that increase never means "off") that does seem to indicate that we want to start with a minimum of 1.

I kind of find this either a failure of the spec for making "0"  and "null" both valid values for the speed setting.

Reading the spec it seems to me that speedSetting MUST NEVER be null, but rather null value is presented as a convenience to the user for "writing null to this value means no change"

Reply:
I would revise proposal as

```cpp
newSpeedSetting = std::invoke([&]{
   VerifyOrReturn(speedSetting.Value() < speedMax, speedMax); 
   return static_cast<uint8_t>(speedSetting.ValueOr(0) + 1);
});

this seems quite readable regarding intent.

Reply:
The speedSetting could be null since the spec said
```
4.4.6.1.2. Auto Value
Setting the attribute value to Auto SHALL set the values of these attributes to null:

PercentSetting
SpeedSetting (if present)
```
So if someone call Step(Increase) when fan mode is **Auto**, I think we can count from `speedCurrent`. If `speedCurrent` is unfortunately null (I didn't see the possibility), I think it's fair to set the speed as 1;

Reply:
Rewrited as below. PTAL
```
+        newSpeedSetting = std::invoke([&]() -> uint8_t {
+            VerifyOrReturnValue(speedValue < speedMax, (aWrap ? kLowestSpeed : speedMax));
+            return static_cast<uint8_t>(speedValue + 1);
+        });

```